geometric_kernels/sampling/feature_maps.py

```diff
@@ -0,0 +1,72 @@
+"""
+Feature maps
+"""
+from typing import Any, Dict, Tuple
+
+import lab as B
+
+from geometric_kernels.kernels import MaternKarhunenLoeveKernel
+from geometric_kernels.lab_extras import from_numpy
+from geometric_kernels.spaces import DiscreteSpectrumSpace, Space
+from geometric_kernels.types import FeatureMap
+
+
+def fixed_feature_map(
+    space: DiscreteSpectrumSpace,
+    kernel: MaternKarhunenLoeveKernel,
+    params,
+    state,
+) -> Tuple[FeatureMap, Dict[str, Any]]:
+    assert "eigenvalues_laplacian" in state
+    assert "eigenfunctions" in state
+
+    repeated_eigenvalues = space.get_repeated_eigenvalues(kernel.num_eigenfunctions)
+    spectrum = kernel._spectrum(
+        repeated_eigenvalues**0.5,
+        nu=params["nu"],
+        lengthscale=params["lengthscale"],
+    )
+
+    weights = B.power(spectrum, 0.5)  # [M, 1]
+    Phi = state["eigenfunctions"]
+
+    def _map(X: B.Numeric) -> B.Numeric:
+        eigenfunctions = Phi.__call__(X, **params)  # [N, M]
+        # return B.matmul(eigenfunctions, weights)  # [N, 1]
+        return eigenfunctions * weights.T  # [N, M]
+
+    _context: Dict[str, str] = {}  # no context
+
+    return _map, _context
+
+
+def random_phase_feature_map(
+    space: Space,
+    kernel: MaternKarhunenLoeveKernel,
+    params,
+    state,
+    key,
+    order=100,
+) -> Tuple[FeatureMap, Dict[str, Any]]:
+
+    key, random_phases = space.random(key, order)  # [O, D]
```

Comment:
We might want to do resampling for each single eigenspace. Will this be too computationally/memory heavy?